src/main/java/io/debezium/connector/vitess/VitessConnector.java

```diff
@@ -20,31 +26,143 @@
+        final Map<String, String> gtidsPerShard = new HashMap<>();
+        for (int i = 0; i < numTasks; i++) {
+            String taskKey = VitessConnector.getTaskKeyName(i, numTasks, gen);
+            VitessPartition par = new VitessPartition(connectorConfig.getLogicalName(), taskKey);
+            Map<String, Object> offset = offsetStorageReader.offset(par.getSourcePartition());
+            if (offset == null && gen == 0) {
+                LOGGER.info("No previous offset for partition: {}, fall back to only server key", par);
+                par = new VitessPartition(connectorConfig.getLogicalName(), null);
+                offset = offsetStorageReader.offset(par.getSourcePartition());
+            }
+            if (offset == null) {
+                if (expectsOffset) {
+                    throw new IllegalArgumentException(String.format("No offset found for %s", par));
+                }
+                else {
+                    LOGGER.warn("No offset found for key key: {}", taskKey);
+                    continue;
+                }
+            }
+            final String vgtidStr = (String) offset.get(SourceInfo.VGTID_KEY);
+            Objects.requireNonNull(vgtidStr, String.format("No vgtid from %s", offset));
+            List<Vgtid.ShardGtid> shardGtids = Vgtid.of(vgtidStr).getShardGtids();
+            for (Vgtid.ShardGtid shardGtid : shardGtids) {
+                gtidsPerShard.put(shardGtid.getShard(), shardGtid.getGtid());
+            }
+        }
+        return gtidsPerShard;
+    }
+
     @Override
     public List<Map<String, String>> taskConfigs(int maxTasks) {
-        if (maxTasks > 1) {
-            throw new IllegalArgumentException("Only a single connector task may be started");
+        LOGGER.info("Calculating taskConfigs for {} tasks", maxTasks);
+        List<String> shards = connectorConfig.offsetStoragePerTask() ? getVitessShards(connectorConfig) : null;
+        return taskConfigs(maxTasks, shards);
+    }
+
+    public List<Map<String, String>> taskConfigs(int maxTasks, List<String> shards) {
+        LOGGER.info("Calculating taskConfigs for {} tasks and shards: {}", maxTasks, shards);
+        if (connectorConfig.offsetStoragePerTask()) {
+            final int prevNumTasks = connectorConfig.getPrevNumTasks();
+            final int gen = connectorConfig.getOffsetStorageTaskKeyGen();
+            Map<String, String> prevGtidsPerShard = gen > 0 ? getGtidPerShardFromStorage(prevNumTasks, gen - 1, true) : null;
+            LOGGER.info("Previous gtids Per shard: {}", prevGtidsPerShard);
+            int tasks = Math.min(maxTasks, shards.size());
+            LOGGER.info("There are {} vitess shards for maxTasks: {}, we will use {} tasks",
+                    shards.size(), maxTasks, tasks);
+            if (gen > 0 && tasks == prevNumTasks) {
+                throw new IllegalArgumentException(String.format(
+                        "Previous num.tasks: %s and current num.tasks: %s are the same",
```

Comment:
Maybe consider adding an explanation and hint for the next step such as "the previous generation must have different number of tasks. Choose a different max.tasks or new shards and retry"?

Reply:
Nice catch, fixed in the new commit.